gloo/cuda_collectives_device.h

```diff
@@ -32,120 +32,228 @@ class CudaLocalDeviceReduce : public LocalOp<T> {
+      if (it->empty()) {
+        it = indices_.erase(it);
+      } else {
+        it++;
+      }
+    }
+    // Best effort alignment so that the group in indices[0] is close to target
+    if (indices_.size() > 1) {
+      alignIndicesWithTarget(groups);
+    }
 
-        // Queue reduction
-        fn_->call(
-            devicePtrs_[indexA],
-            devicePtrs_[indexB],
-            devicePtrs_[indexA].getCount(),
-            streamA);
+    // Shuffle order in an attempt to evenly spread work across devices when
+    // dealing with multiple instances of this operation.
+    auto& groupA = indices_[0];
+    std::random_shuffle(groupA.begin(), groupA.end());
+
+    // Make sure that the devices associated with the first element in
+    // each vector have peer to peer access, so we can reduce between groups.
+    // This is possible if this host uses the canonical NVLink topology.
+    if (indices_.size() > 1) {
+      auto devA = devicePtrs_[groupA.front()].getDeviceID();
+      auto& groupB = indices_[1];
+
+      // Find device with peer access to devA
+      auto ok = false;
+      for (auto it = groupB.begin(); it != groupB.end(); it++) {
+        if (canAccessPeer(devA, devicePtrs_[*it].getDeviceID())) {
+          // Move this index to front of vector
+          auto index = *it;
+          groupB.erase(it);
+          groupB.insert(groupB.begin(), index);
+          // Shuffle the remainder
+          std::random_shuffle(groupB.begin() + 1, groupB.end());
+          ok = true;
+          break;
+        }
       }
+      GLOO_ENFORCE(ok, "Expected to find peer access between groups");
     }
+  }
 
-    // Queue copy to target on the root stream
-    auto root = indices_[0];
-    streams_[root].copyAsync(targetPtr_, devicePtrs_[root]);
+  void enablePeerAccess(int indexA, int indexB) {
+    auto devA = devicePtrs_[indexA].getDeviceID();
+    auto devB = devicePtrs_[indexB].getDeviceID();
+
+    // Number of elements must be equal
+    GLOO_ENFORCE_EQ(
+        devicePtrs_[indexA].getCount(),
+        devicePtrs_[indexB].getCount());
+
+    // Devices must be able to access each others memory
+    GLOO_ENFORCE_EQ(
+        1,
+        canAccessPeer(devA, devB),
+        "GPU ",
+        devA,
+        " does not have peer access to GPU ",
+        devB);
+
+    // Enable peer access for devA to memory on devB
+    CudaDeviceScope scope(devA);
+    cudaDeviceEnablePeerAccess(devB, 0);
+
+    // Use cudaGetLastError so that any error is cleared.
+    auto err = cudaGetLastError();
+    if (err != cudaErrorPeerAccessAlreadyEnabled) {
+      CUDA_CHECK(err);
+    }
   }
 
-  virtual void wait() {
-    // Wait for the final memory copy to complete
-    auto root = indices_[0];
-    streams_[root].wait();
+  void initializePeerAccess() {
+    // Initialize peer access within groups
+    for (const auto& group : indices_) {
+      int steps = log2(group.size());
+      GLOO_ENFORCE(
+          1 << steps,
+          group.size(),
```

Comment:
one thing I dont see being taken into account is that peer access has a device limit of 8.
Even if deviceA and deviceB can be p2p enabled, after you register 8 devices for p2p, the subsequent `cudaDeviceEnablePeerAccess` calls will start getting rejected.
We particularly see this issue on 8 x K80 machines which see 16 visible devices.

Both Caffe2 and Torch work around this. See the Torch patch for example: https://github.com/torch/cutorch/pull/774